src/texture-processor/detail/iterator.hh

```diff
@@ -0,0 +1,169 @@
+#pragma once
+
+#include <clean-core/always_false.hh>
+#include <clean-core/sentinel.hh>
+
+#include <typed-geometry/tg-lean.hh>
+
+#include <texture-processor/detail/predicates.hh>
+
+namespace tp
+{
+template <int D, class PixelT>
+struct pixel_entry
+{
+    tg::pos<D, int> const pos;
+    PixelT pixel;
+};
+}
+
+namespace tp::detail
+{
+template <class iterator_t>
+struct srange
+{
+    iterator_t it;
+    srange(iterator_t it) : it(it) {}
+
+    iterator_t begin() const { return it; }
+    cc::sentinel end() const { return {}; }
+};
+
+// TODO: can probably be optimized further
+template <int D>
+struct strided_linear_pos_iterator
```

Comment:
This is basically a linear-to-multidimensional indexing helper with support for negative strides, right? I think basic free-function variants going both ways would help a lot, with this iterator using them.